### PR TITLE
3d curve tangent estimator

### PR DESCRIPTION
**Pull request by JacquesOlivierLachaud**
A simple tool to estimate the tangent to a set of points approaching a 3D curve, based on digital Voronoi Covariance Measure.

### CODE REVIEW

**Comment by copyme** [comment]
Hi! For option -2 I got:
 [ERR] Assertion Error - assertion (myCellEmbedder != 0) failed in DGtal::Display3D<Space, KSpace>::RealPoint DGtal::Display3D<Space, KSpace>::embedK(const typename KSpace::Cell&) const [with Space = DGtal::SpaceND<3u, int>; KSpace = DGtal::KhalimskySpaceND<3u, int>; DGtal::Display3D<Space, KSpace>::RealPoint = DGtal::PointVector<3u, double>; typename KSpace::Cell = DGtal::KhalimskyCell<3u, int>]: /usr/local/include/DGtal/io/Display3D.ih(1220)

**Comment by JacquesOlivierLachaud** [comment]
It is always the same error. The viewer needs an instance of the Khalimsky space.

**Comment by copyme** [comment]
Projections sometimes do not contain all points 
missing-points 

Presented curve can be downloaded from: https://www.wuala.com/hilsim/Bugs/DGTal/?key=ghsyUCKCLtzp

Moreover, I noticed that program randomly starts fast and another time I need to wait really long to see the viewer window. I will try to debug.

**Comment by JacquesOlivierLachaud** [comment]
Thanks for debugging. Kind of weird bug. The problem is maybe that I still use the GridCurve at some point, whereas a GridCurve requires face adjacency.

**Comment by copyme** [comment]
I realized that that bug with the missing points was related to a change in your code which I did in a mean time -- I've replaced Standard6DSSComputer by my Naive3DDSSComputer (because of that I removed my previous comment). Nevertheless, I do not have any clue why run time is different randomly, but I will definitely take a look on this issue. Yes, maybe this is because of GridCurve face adjacency.

**Comment by copyme** [comment]
@JacquesOlivierLachaud Seems that it is viewer << Viewer3D<Z3,K3>::updateDisplay; which takes sometimes lot of time. No clue yet why. Well, I've added LMST to your code. If you agree to add LMST to DGTal right now, I can request merge && pull. Let me know what do you think.

**Comment by JacquesOlivierLachaud** [comment]
If it is updateDisplay, I think it is QGLviewer related. 
For LMST, this is excellent news. Of course, I agree with a PR. 
Jaco 

----- Mail original -----

De: "kacperp" <notifications@github.com> 
À: "DGtal-team/DGtalTools" <DGtalTools@noreply.github.com> 
Cc: "Jacques-Olivier Lachaud" <jacques-olivier.lachaud@univ-savoie.fr> 
Envoyé: Mardi 23 Juin 2015 13:12:20 
Objet: Re: [DGtalTools] 3d curve tangent estimator (#218) 



@JacquesOlivierLachaud Seems that it is viewer << Viewer3D::updateDisplay; which takes sometimes lot of time. No clue yet why. Well, I've added LMST to your code. If you agree to add LMST to DGTal right now, I can request merge && pull. Let me know what do you think.

**Comment by copyme** [comment]
@JacquesOlivierLachaud Done, moreover modified version of 3dCurvetTangentEstimator you can find here: https://github.com/DGtal-team/DGtalTools/blob/3DCurvatureViewer/estimators/3dCurveTangentEstimator.cpp

**Comment by JacquesOlivierLachaud** [comment]
I think you've got compiling errors because of some C++11 features (like using cbegin and cend ). Perhaps you should use them only when the flag CPP11 is on.

**Comment by JacquesOlivierLachaud** [comment]
Oups, the previous comment was for another PR.

**Comment by copyme** [comment]
Is this still active? Do we want to add this tool to DGtalTools 0.9?

**Comment by dcoeurjo** [comment]
agree with @copyme , should this tool be included in the 0.9 release ?

**Comment by copyme** [comment]
@dcoeurjo I made a pull & merge request for @JacquesOlivierLachaud with my branch where I added a support for the LMST.

**Comment by JacquesOlivierLachaud** [comment]
Sorry. I missed something. Is this PR now included into the lMST PR ?

**Comment by copyme** [comment]
@JacquesOlivierLachaud Up to you. But because there is not that much of time before release of 0.9 and seems that LMST will appear in it I made a PR for you. Agree that maybe it is not the best solution. Sorry for the confusion.

**Comment by JacquesOlivierLachaud** [comment]
Hello kacper
I've just understood what you mean ;-P
Ok, I look at the PR in my repo first then I PR on DGtal-team.
Jaco
PS: I must first finish a review this morning then I take care of it.

**Comment by JacquesOlivierLachaud** [comment]
I think this PR is ready for merge (waiting for travis).

File: CMakeLists.txt [hunk 1]
```diff
@@ -38,7 +38,7 @@ SET(DGTALTOOLS_VERSION "${DGtalTools_VERSION_MAJOR}.${DGtalTools_VERSION_MINOR}.
 SET(PROJECT_VERSION "${DGtalTools_VERSION_MAJOR}.${DGtalTools_VERSION_MINOR}.${DGtalTools_VERSION_PATCH}")
 SET(VERSION ${DGtalTools_VERSION_MAJOR}.${DGtalTools_VERSION_MINOR}.${DGtalTools_VERSION_PATCH})
 
-SET(Boost_USE_STATIC_LIBS   ON)
+SET(Boost_USE_STATIC_LIBS   OFF)
```
Comment (dcoeurjo) [inline CMakeLists.txt]: why did you need to switch this off ? 
Static libs help in MSWindows.

File: ChangeLog [hunk 2]
```diff
@@ -22,6 +22,9 @@
 	- heightfield2shading: a new tool to render a 2D heightfield image into a shading one.
 	- mesh2heightfield:  new tool to convert a mesh file into a 2D heightmap (from a normal direction N and from a starting point P).
 
+* estimators/:
+	- 3dCurveTangentEstimator: a simple tool to estimate and visualize the tangent to a set of points approaching a 3D curve. Two estimators are implemented, one based on digital Voronoi Covariance Measure, the other based on 3D lambda-MST.
```
Comment (dcoeurjo) [inline ChangeLog]: approaching -> approximating?

Comment (dcoeurjo) [inline ChangeLog]: keep it 80col ?

File: estimators/3dCurveTangentEstimator.cpp [hunk 3]
```diff
@@ -0,0 +1,778 @@
...
+      std::vector<PointD2d> pts2d;
+      pts2d.push_back( dss2d.project(dss2d.back(), dss2d.Uf()) );
+      pts2d.push_back( dss2d.project(dss2d.back(), dss2d.Lf()) );
+      pts2d.push_back( dss2d.project(dss2d.front(), dss2d.Lf()) );
+      pts2d.push_back( dss2d.project(dss2d.front(), dss2d.Uf()) );
+      std::vector<PointD3D> bb;
+      PointD3D p3;
+      for ( unsigned int j = 0; j < pts2d.size(); ++j )
+  {
+    switch (i) {
```
Comment (dcoeurjo) [inline estimators/3dCurveTangentEstimator.cpp]: coding style: line break before ```{```

File: CMakeLists.txt (hunk 1 above)
Comment (copyme) [inline CMakeLists.txt]: @dcoeurjo Sorry, this should not be committed. On my local configuration I do not have a static boost.

File: estimators/3dCurveTangentEstimator.cpp [hunk 4]
```diff
@@ -0,0 +1,778 @@
...
+    po::store( clp.run(), vm );
+  } catch( const exception& ex ) {
+    parseOK = false;
+    trace.info() << "Error checking program options: "<< ex.what() << endl;
+  }
+  po::notify( vm );
+  if( !parseOK || vm.count("help")||argc<=1)
+  {
+    cout << "Usage: " << argv[0] << " [options] input\n"
+    << "This program estimates the tangent vector to a set of 3D integer points, which are supposed to approach a 3D curve. This set of points is given as a list of points in file <input>. The tangent estimator uses either the digital Voronoi Covariance Measure (VCM) or the 3D lambda-Maximal Segment Tangent (L-MST). This program can also displays the curve and tangent estimations, and it can also extract maximal digital straight segments (2D and 3D)."
```
Comment (dcoeurjo) [inline estimators/3dCurveTangentEstimator.cpp]: approach->approximate

File: visualisation/3dCurveViewer.cpp [hunk 5]
```diff
@@ -153,11 +153,24 @@ void displayDSS3dTangent( Viewer3D<space, kspace> & viewer,
...
   typedef typename Display3D<>::BallD3D PointD3D;
   Point directionZ3;
-  PointD3d P1, P2, direction, intercept, thickness;
-  dss3d.getParameters( directionZ3, intercept, thickness );
+  PointR3d interceptR, thicknessR;
+  PointD3d P1, P2, direction;
+  dss3d.getParameters( directionZ3, interceptR, thicknessR );
+  
+  PointD3d intercept;
+  intercept[0] = (double) NumberTraits<int>::castToInt64_t ( interceptR[0].first ) / (double) NumberTraits<int>::castToInt64_t ( interceptR[0].second );
```
Comment (dcoeurjo) [inline visualisation/3dCurveViewer.cpp]: why not doing a castToDouble instead of ```(double)castToInt64_t```?

Comment (copyme) [inline visualisation/3dCurveViewer.cpp]: @dcoeurjo Agree, this could be also done in two places in DGtal code. Should I do it?

Comment (dcoeurjo) [inline visualisation/3dCurveViewer.cpp]: please do.

Comment (copyme) [inline visualisation/3dCurveViewer.cpp]: @dcoeurjo I am not sure if @JacquesOlivierLachaud can fix your comments today. But If you merge his code I can fix it after in a new RP.

Comment (dcoeurjo) [inline visualisation/3dCurveViewer.cpp]: Ok. Could you please also fix the other issues ?

Comment (copyme) [inline visualisation/3dCurveViewer.cpp]: Sure.

**Comment by dcoeurjo** [comment]
Merging 
(Ping @copyme)
